src/test/regress/sql/local_table_join.sql

```diff
@@ -0,0 +1,316 @@
+
+SELECT count(*) FROM postgres_table JOIN distributed_table_pkey ON postgres_table.key = 10;
+
+
+select count(*) FROM postgres_table JOIN (SELECT a.key,random() FROM distributed_table a JOIN distributed_table b USING(key)) as foo USING(key);
+select count(*) FROM (SELECT a.key, random() FROM distributed_table a JOIN distributed_table b USING(key)) as foo JOIN postgres_table  USING(key);
+
+SELECT count(*) FROM postgres_table JOIN (SELECT * FROM distributed_table) d1 USING(key);
+-- since this is already router plannable, we don't recursively plan the postgres table
+SELECT count(*) FROM postgres_table JOIN (SELECT * FROM distributed_table LIMIT 1) d1 USING(key);
+
+-- a unique index on key so dist table should be recursively planned
+SELECT count(*) FROM postgres_table JOIN distributed_table_windex USING(key);
+SELECT count(*) FROM postgres_table JOIN distributed_table_windex USING(value);
+SELECT count(*) FROM postgres_table JOIN distributed_table_windex ON postgres_table.key = distributed_table_windex.key;
+SELECT count(*) FROM postgres_table JOIN distributed_table_windex ON distributed_table_windex.key = 10;
+
+-- no unique index on value so local table should be recursively planned.
+SELECT count(*) FROM distributed_table JOIN postgres_table USING(key) WHERE distributed_table.value = 'test';
+
+SELECT count(*) FROM distributed_table JOIN postgres_table USING(key) WHERE distributed_table.key = 1;
+
+
+-- if both local and distributed tables have a filter, we prefer local unless distributed table has unique indexes on any equality filter
+SELECT count(*) FROM distributed_table JOIN postgres_table USING(key) WHERE distributed_table.value = 'test' AND postgres_table.value = 'test';
+SELECT count(*) FROM distributed_table JOIN postgres_table USING(key) WHERE distributed_table.value = 'test' OR postgres_table.value = 'test';
+
+
+-- multiple local/distributed tables
+-- only local tables are recursively planned
+SELECT count(*) FROM distributed_table d1 JOIN postgres_table p1 USING(key) JOIN distributed_table d2 USING(key) JOIN postgres_table p2 USING(key);
+
+
+SELECT
+	count(*)
+FROM
+	distributed_table d1 JOIN postgres_table p1 USING(key) JOIN distributed_table d2 USING(key) JOIN postgres_table p2 USING(key)
+WHERE
+	d1.value = '1';
+
+-- if the filter is on the JOIN key, we can recursively plan the local
+-- tables as filters are pushed down to the local tables
+SELECT
+	count(*)
+FROM
+	distributed_table d1 JOIN postgres_table p1 USING(key) JOIN distributed_table d2 USING(key) JOIN postgres_table p2 USING(key)
+WHERE
+	d1.key = 1;
+
+CREATE view loc_view AS SELECT * FROM postgres_table WHERE key > 0;
+UPDATE loc_view SET key = (SELECT COUNT(*) FROM distributed_table);
+
+SELECT count(*)
+FROM
+	(SELECT * FROM (SELECT * FROM distributed_table) d1) d2
+JOIN postgres_table
+USING(key);
+
+
+-- TODO:: we should support this?
```

Comment:
any update?

Reply:
here it still doesn't work because reference tables are not considered as `multi shard query`. Why is this `multi shard query` but not `multi placement query`? I am specifically talking about `isMultiShardQuery` flag in `PlanRouterQuery`